[`flake8-pyi`] Implement autofix for `redundant-numeric-union` (`PYI041`)

sbrugman opened the pull request:
## Summary

This PR adds autofix for `redundant-numeric-union` (`PYI041`)

There are some comments below to explain the reasoning behind some choices that might help review.

<!-- What's the purpose of the change? What does it do, and why? -->

Resolves part of https://github.com/astral-sh/ruff/issues/14185.

## Test Plan

<!-- How was it tested? -->

sbrugman commented on crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:
```diff
@@ -47,16 +61,21 @@ pub struct RedundantNumericUnion {
     #[derive_message_formats]
     fn message(&self) -> String {
         let (subtype, supertype) = match self.redundancy {
+            Redundancy::IntFloatComplex => ("int | float", "complex"),
```
Emitting multiple messages for the same union, such as "Use `complex` instead of `int | complex`" and  "Use `complex` instead of `float | complex`" gets confusing when there is a fix presented:

```diff
-    def bad4(self, arg: Union[float | complex, int]) -> None: ...  # PYI041
+    def bad4(self, arg: complex) -> None: ...  # PYI041
```

With `IntFloatComplex` we can assign an unique violation for each combination.

sbrugman commented on crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:
```diff
@@ -66,57 +85,212 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
-}
-
-fn check_annotation(checker: &mut Checker, annotation: &Expr) {
-    let mut has_float = false;
```
Replace `has_*` with `bitflags`, as they correspond naturally to the order of Complex > Float > Int.

sbrugman commented on crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:
```diff
@@ -66,57 +85,212 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
     };
 
+    // Traverse the union, and remember which numeric types are found.
     traverse_union(&mut find_numeric_type, checker.semantic(), annotation);
```
The second loop is only active when there is a violation, instead of collecting all expressions for any Union and filtering them later to avoid overhead in those cases.

sbrugman commented on crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:
```diff
@@ -66,57 +85,212 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
+    pub(super) fn from_numeric_flags(numeric_flags: NumericFlags) -> Option<Self> {
+        match numeric_flags.bits() {
+            // NumericFlags::INT | NumericFlags::FLOAT | NumericFlags::COMPLEX
+            0b0111 => Some(Self::IntFloatComplex),
```
There must be a more elegant way of doing this, but I haven't found it yet and this works.

MichaReiser replied:
I don't think there is and it's inherent to bitflags where it isn't required that all possible combinations are valid. 

I would write this as

```
        if numeric_flags == NumericFlags::INT | NumericFlags::FLOAT | NumericFlags::COMPLEX {
            Some(Self::IntFloatComplex)
        } else if numeric_flags == NumericFlags::INT | NumericFlags::FLOAT {
            Some(Self::IntComplex)
        } else if numeric_flags == NumericFlags::FLOAT | NumericFlags::COMPLEX {
            Some(Self::FloatComplex)
        } else if numeric_flags == NumericFlags::FLOAT | NumericFlags::INT {
            Some(Self::IntFloat)
        } else {
            None
        }
```

sbrugman replied:
Thanks, yes that looks better.

sbrugman commented on crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:
```diff
@@ -66,57 +85,212 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
+    annotation: &Expr,
+    applicability: Applicability,
+) -> Fix {
+    debug_assert!(nodes.len() >= 2, "At least two nodes required");
```
The code above handles `nodes.len() < 2` separately and thus this function should never encounter these cases. Is `debug_assert!` ok to denote this?

sbrugman commented on crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:
```diff
@@ -66,57 +85,212 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
+        Applicability::Safe
+    };
+
+    // Generate the flattened fix once.
```
TODO(SB):     
```rs
if checker.settings.preview.is_enabled() {
```

MichaReiser commented on crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:
````diff
@@ -37,6 +44,13 @@ use crate::checkers::ast::Checker;
 /// ```
 ///
+/// ## Fix safety
+/// This rule's fix is marked as safe for most cases; however, the fix will
+/// nested unions type expressions into a single top-level union.
````
the fix will...?

MichaReiser commented on crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:
```diff
@@ -37,6 +44,13 @@ use crate::checkers::ast::Checker;
+/// This rule's fix is marked as safe for most cases; however, the fix will
+/// nested unions type expressions into a single top-level union.
+///
+/// The fix is marked as unsafe when comments are present within the type
+/// expression.
```
What's the reason that the fix is marked as unsafe when comments are present? Does it produce invalid syntax? Are there cases it can't handle?

I think our general practice has now been to not mark a fix as unsafe just because it removes comments. Instead, we stated that the fix could remove comments.

MichaReiser commented on crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:
```diff
@@ -66,57 +88,214 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
+        let fix = if let &[edit_expr] = necessary_nodes.as_slice() {
+            // Generate a [`Fix`] for a single type expression, e.g. `int`.
+            Fix::applicable_edit(
+                Edit::range_replacement(checker.generator().expr(edit_expr), annotation.range()),
```
Is there a specific reason for using a generator over a text-based fix? Most fixes are text based to preserve existing formatting (and possibly comments).

sbrugman replied:
I've used the generator based on other rules, it could very well be that a text-based fix is more optimal.

sbrugman replied:
However, redundant expressions are still dropped. If they have the comment, it will be still lost.

MichaReiser replied:
> I've used the generator based on other rules, it could very well be that a text-based fix is more optimal.

We generally prefer using text-based fixes if they are not too complicated because they preserve more comments and the formatting. But I don't think it's worth rewriting it now, or we can do so as a separate PR .

MichaReiser commented on crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:
```diff
@@ -66,57 +88,214 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
+        // `int` is always dropped, since `float` or `complex` must be present.
+        // `float` is only dropped if `complex`` is present.
+        if (builtin_type == "float" && !numeric_flags.contains(NumericFlags::COMPLEX))
+            || (builtin_type != "float" && builtin_type != "int")
```
Isn't this the same as?

```suggestion
            || builtin_type == "complex"
```

Nit: It could also be worth to have a `BuiltInTy` enum with `Float`, `Int`, and `Complex` to reduce some of the string matching

sbrugman replied:
Here I think the `builtin_type` is not guaranteed to be `int`/`float`/`complex`, in which case we cannot replace this condition

MichaReiser replied:
Oh, in case it is e.g. `str` or `bool` and I think there are unit tests that exercise this.

MichaReiser commented on crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:
```diff
@@ -66,57 +88,214 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
+        }
+    };
+
+    // Traverse the union, and remember which numeric types are found.
```
This comment seems a copy-paste leftover from above. We aren't traversing the union to remember which numeric types are found, but to generate the fix.

MichaReiser commented on crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:
```diff
@@ -66,57 +88,214 @@ pub(crate) fn redundant_numeric_union(checker: &mut Checker, parameters: &Parame
+            ctx: ExprContext::Store,
+            range: TextRange::default(),
+        })),
+        slice: Box::new(if let [elt] = nodes.as_slice() {
+            (*elt).clone()
+        } else {
```
Isn't this branch dead because of the assertion above?

sbrugman replied:
Good catch!

MichaReiser reviewed (commented):
Thanks. This overall looks great. 

The main open questions are

* should use text based edits to avoid reformatting unchanged code and loosing inner comments. 
* Whether it's possible to always mark the fix as safe, regardless of comments

MichaReiser commented:
Oh, and did you look through the ecosystem changes? If so, could you update the test plan. It's helpful to know if you did because I then don't need to carefully review all of them and can instead just click through some of them.

sbrugman commented:
> Oh, and did you look through the ecosystem changes? If so, could you update the test plan. It's helpful to know if you did because I then don't need to carefully review all of them and can instead just click through some of them.

The ecosystem results look good.
From [this example](https://github.com/apache/superset/blob/0e165c1a21a90098adfea8efa5006e4feb2adf11/superset/utils/core.py#L349) I'm wondering if we shouldn't include return types; it appears to skip that one now.

Typeshed shows negatives in the ecosystem results, but I expect that to not be related to the PR. Maybe @AlexWaygood can help out with that.

MichaReiser commented:
> From [this example](https://github.com/apache/superset/blob/0e165c1a21a90098adfea8efa5006e4feb2adf11/superset/utils/core.py#L349) I'm wondering if we shouldn't include return types; it appears to skip that one now.

Could you expand what you mean by *it appears to skip that now*. Is it not flagging this case or do you think this case shouldn't be flagged?

MichaReiser commented on crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:
````diff
@@ -37,6 +44,13 @@ use crate::checkers::ast::Checker;
 /// ```
 ///
+/// ## Fix safety
+/// This rule's fix is marked as safe for most cases; however, the fix will
+/// flatten nested unions type expressions into a single top-level union.
````
Reading the documentation, I get the impression the fix isn't safe because it flattens nested union expressions. But I understand that this doesn't make the fix unsafe, it's just an extension where the fix goes beyond what one might expect.

sbrugman replied:
Exactly. I've reused the phrasing that Charlie used for another rule.

The nesting is a feature to allow cases such as these:
```python
ReadOnlyMode         = Literal["r", "r+"]
WriteAndTruncateMode = Literal["w", "w+", "wt", "w+t"]
WriteNoTruncateMode  = Literal["r+", "r+t"]
AppendMode           = Literal["a", "a+", "at", "a+t"]

AllModes = Literal[ReadOnlyMode, WriteAndTruncateMode,
                   WriteNoTruncateMode, AppendMode]
```

As a consequence this redundant nesting is also allowed, but can be safely fixed:

```python
AllModes = Literal[Literal["r", "r+"], Literal["w", "w+", "wt", "w+t"],
                   Literal["r+", "r+t"], Literal["a", "a+", "at", "a+t"]]
```

See https://typing.readthedocs.io/en/latest/spec/literal.html#legal-and-illegal-parameterizations

We probably should have a rule for this (for both `union` and `literal`).

charliermarsh replied:
I tweaked the comment a bit on all three rules.

MichaReiser commented on crates/ruff_linter/src/rules/flake8_pyi/rules/redundant_numeric_union.rs:
```diff
@@ -48,15 +62,23 @@ pub struct RedundantNumericUnion {
     }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Remove duplicates".to_string())
```
Nit: To align with the terminology used in the documentation

```suggestion
        Some("Remove redundant type".to_string())
```

AlexWaygood commented:
> Typeshed shows negative in the ecosystem results, but I expect that to not be related to the PR.
> ~Maybe @AlexWaygood can help out with that.~
> It's probably because the example is Py3.13+

The ecosystem reports for typeshed have [often not been reproducible locally](https://github.com/astral-sh/ruff/issues/11305) and nobody's yet had a chance to look into why... if you can't repro the typeshed hits locally, I'd just ignore them!

sbrugman commented:
> Could you expand what you mean by it appears to skip that now. Is it not flagging this case or do you think this case shouldn't be flagged?

After investigating it turned out that this rule hooked into `statement.rs` instead of `expression.rs` like all the other rules related to unions do. This caused false negatives on the return types and type aliases. We had no test for this. Fixed it, and added a test.

MichaReiser commented on crates/ruff_linter/src/checkers/ast/analyze/expression.rs:
```diff
@@ -96,6 +97,9 @@ pub(crate) fn expression(expr: &Expr, checker: &mut Checker) {
                     if checker.enabled(Rule::UnnecessaryTypeUnion) {
                         flake8_pyi::rules::unnecessary_type_union(checker, expr);
                     }
+                    if checker.enabled(Rule::RedundantNumericUnion) {
+                        flake8_pyi::rules::redundant_numeric_union(checker, expr);
+                    }
```
Oh nice find. I think we should move this change out into its own PR so that it gets its own entry in the changelog. I have to think about it if it warrants a breaking change or not.

sbrugman replied:
I'll move the last commit into it's own PR
